docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,551 @@
+customization than simple key/value pairs.
+
+As an example, we may define a GVK that contains service endpoints that many of
+applications use. In each application package, we would then include an instance
+of that resource, say called "service-endpoints", and configure a function to
+propagate the values from that resource to others within our package. As those
+endpoints may vary by region, in our Porch cluster we can create an instance of
+this GVK for each region: "useast1-service-endpoints",
+"useast2-service-endpoints", "uswest1-service-endpoints", etc. When we
+instantiate the PackageVariant for a cluster, we want to inject the resource
+corresponding to the region in which the cluster exists. Thus, for each cluster
+we will create a PackageVariant resource pointing to the upstream package, but
+with injection selector values that are specific to the region for that cluster.
+
+It is important to realize that the name of the in-package resource and the in-
+cluster resource need not match. In fact, it would be an unusual coincidence if
+they did match. The names in the package are the same across PackageVariants
+using that upstream, but we want to inject different resources for each one such
+PackageVariant. We also do not want to change the name in the package, because
+it likely has meaning within the package and will be used by functions in the
+package. Also, different roles controlling the names of the in-package and in-
+cluster resources. The names in the package are in the control of the package
+author. The names in the cluster are in the control of whoever populates the
+cluster (for example, some infrastructure team). The selector is the glue
+between them, and is in control of the PackageVariant resource creator.
+
+The GVK on the other hand, has to be the same for the in-package resource and
+the in-cluster resource, because it tells us the API schema for the resource.
+Also, the namespace of the in-cluster object needs to be the same as the
+PackageVariant resource, or we could leak resources from namespaces to which
+our PackageVariant user does not have access.
+
+With that understanding, the injection process works as follows:
+
+1. The controller will examine all in-package resources, looking for those with
+   an annotation named `kpt.dev/config-injection`, with one of the following
+   values: `required` or `optional`. We will call these "injection points".
+1. For each injection point, a condition will be created in the
+   downstream PackageRevision, with ConditionType set to the dot-delimited
+   concatenation of `config.injection`, with the in-package resource kind and
+   name, and the value set to `False`. Note that since the package author
+   controls the name of the resource, kind and name are sufficient to
+   disambiguate the injection point. We will call this ConditionType the
+   "injection point ConditionType".
+1. For each required injection point, the injection point ConditionType will be
```

Comment:
So if I understand it correctly, it is the responsibility of the package author to add the required injection points to the PackageRevision. I think it might be useful to clarify it here.

Reply:
I added some clarifying language, let me know if it's still an issue (after I push).